File: backend/src/main/java/com/garagelog/entity/Mantenimiento.java
package com.garagelog.entity;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

@Entity
@Table(name = "mantenimiento")
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Mantenimiento {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_mantenimiento")
    private Integer idMantenimiento;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id_vehiculo")
    @JsonIgnoreProperties("mantenimientos")
    private Vehiculo vehiculo;

    @Column(nullable = false)
    private LocalDate fecha;

    private Integer km;

    @Column(name = "coste_total", precision = 10, scale = 2)
    private BigDecimal costeTotal;

    @Column(columnDefinition = "TEXT")
    private String observaciones;

    @OneToMany(mappedBy = "mantenimiento", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonIgnoreProperties("mantenimiento")
    private List<MantenimientoPieza> piezas;

    public Integer getIdMantenimiento() { return idMantenimiento; }
    public void setIdMantenimiento(Integer idMantenimiento) { this.idMantenimiento = idMantenimiento; }
    public Vehiculo getVehiculo() { return vehiculo; }
    public void setVehiculo(Vehiculo vehiculo) { this.vehiculo = vehiculo; }
    public LocalDate getFecha() { return fecha; }
    public void setFecha(LocalDate fecha) { this.fecha = fecha; }
    public Integer getKm() { return km; }
    public void setKm(Integer km) { this.km = km; }
    public BigDecimal getCosteTotal() { return costeTotal; }
    public void setCosteTotal(BigDecimal costeTotal) { this.costeTotal = costeTotal; }
    public String getObservaciones() { return observaciones; }
    public void setObservaciones(String observaciones) { this.observaciones = observaciones; }
    public List<MantenimientoPieza> getPiezas() { return piezas; }
    public void setPiezas(List<MantenimientoPieza> piezas) { this.piezas = piezas; }
}

File: garage_log/lib/models/mantenimiento.dart
import 'mantenimiento_pieza.dart';

class Mantenimiento {
  int idMantenimiento;
  String fecha;
  int? km;
  double? costeTotal;
  String? observaciones;
  List<MantenimientoPieza>? piezas;

  Mantenimiento({
    required this.idMantenimiento,
    required this.fecha,
    this.km,
    this.costeTotal,
    this.observaciones,
    this.piezas,
  });

  factory Mantenimiento.fromJson(Map<String, dynamic> json) =>
      Mantenimiento(
        idMantenimiento: json["idMantenimiento"],
        fecha: json["fecha"] ?? "",
        km: json["km"],
        costeTotal: (json["costeTotal"] as num?)?.toDouble(),
        observaciones: json["observaciones"],
        piezas: json["piezas"] != null
            ? List<MantenimientoPieza>.from(
                json["piezas"].map((x) => MantenimientoPieza.fromJson(x)))
            : null,
      );

  Map<String, dynamic> toJson() => {
        "idMantenimiento": idMantenimiento,
        "fecha": fecha,
        "km": km,
        "costeTotal": costeTotal,
        "observaciones": observaciones,
        "piezas": piezas?.map((x) => x.toJson()).toList(),
      };
}
